src/librustc/traits/query/normalize.rs

```diff
@@ -81,6 +81,10 @@ impl<'cx, 'tcx> TypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
     }
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        if !ty.has_projections() {
```

Comment:
You should split this out to a separate PR since any performance changes from this PR will be entirely hidden by this performance improvement.

Reply:
Done https://github.com/rust-lang/rust/pull/68606